packages/react-reconciler/src/ReactFiberIncompleteWork.js

```diff
@@ -0,0 +1,387 @@
+      case Err: {
+        const errorInfo = effectValue;
+        errors.push(errorInfo);
+        return true;
+      }
+      default:
+        return false;
+    }
+  }
+
+  function exitIncompleteWork(
+    current: Fiber | null,
+    workInProgress: Fiber,
+    renderIsExpired: boolean,
+    renderExpirationTime: ExpirationTime,
+  ) {
+    let next = null;
+    switch (workInProgress.tag) {
+      case FunctionalComponent:
+        break;
+      case ClassComponent: {
+        popLegacyContextProvider(workInProgress);
+        const ctor = workInProgress.type;
+        const instance = workInProgress.stateNode;
+        if (
+          (workInProgress.effectTag & DidCapture) === NoEffect &&
+          ((typeof ctor.getDerivedStateFromCatch === 'function' &&
+            enableGetDerivedStateFromCatch) ||
+            (instance !== null &&
+              typeof instance.componentDidCatch === 'function' &&
+              !isAlreadyFailedLegacyErrorBoundary(instance)))
+        ) {
+          const errors = [];
+          let previousEffect = null;
+          let effect = workInProgress.firstEffect;
+          while (effect !== null) {
+            const algebraicEffectTag = effect.effectTag & AlgebraicEffectMask;
+            const effectValue = effect.thrownValue;
+            const didHandle = handleClassEffect(
+              workInProgress,
+              errors,
+              algebraicEffectTag,
+              effectValue,
+            );
+            if (didHandle) {
+              captureEffect(workInProgress, previousEffect, effect);
+            }
+            previousEffect = effect;
+            effect = effect.nextEffect;
+          }
+          if (errors.length > 0) {
+            ensureUpdateQueues(workInProgress);
+            const updateQueue: UpdateQueue<
+              any,
+            > = (workInProgress.updateQueue: any);
+            updateQueue.capturedValues = errors;
+            workInProgress.effectTag |= DidCapture;
+            // Render the boundary again
+            next = workInProgress;
+          }
+        }
+        break;
+      }
+      case HostRoot:
+        popHostContainer(workInProgress);
+        popTopLevelLegacyContextObject(workInProgress);
+        next = exitIncompleteRoot(
+          workInProgress,
+          renderIsExpired,
+          renderExpirationTime,
+        );
+        break;
+      case HostComponent:
+        popHostContext(workInProgress);
+        break;
+      case HostText:
+        break;
+      case CallComponent:
+        break;
+      case CallHandlerPhase:
+        break;
+      case ReturnComponent:
+        break;
+      case Fragment:
+        break;
+      case Mode:
+        break;
+      case HostPortal:
+        popHostContainer(workInProgress);
+        break;
+      case ContextProvider:
+        popProvider(workInProgress);
+        break;
+      case ContextConsumer:
+        break;
+      case IndeterminateComponent:
+        break;
+      default:
```

Comment:
Is this code path commonly exercised in tests? Is there a risk that we'll add a new tag, forget to add it to this switch, and it blows up only on the error path (which is rare enough that we will miss it)?

Reply:
It’s triggered whenever anything throws, so we need test cases that throw for every possible type. Working on adding this to the fuzz tester.

Reply:
Seems like a good candidate for a Flow exhaustiveness check.

Reply:
True. Though we still should test every path to prevent stack push/pop mismatch bugs, like we’ve seen in the past.